File: apps/node/protocol_conformance_test.go
// Node-side leg of the tri-implementation Bridge Protocol conformance suite
// (see protocol/README.md). The daemon asserts the same fixtures from
// TypeScript and the companion app from Dart; this leg proves talon-node's
// real code paths — capability advertisement, registration body, SSE frame
// filtering, and command dispatch — agree with the shared wire samples.
package main

import (
	"context"
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"strings"
	"testing"
)

type fixtureCommand struct {
	ID       string         `json:"id"`
	DeviceID string         `json:"deviceId"`
	Name     string         `json:"name"`
	Params   map[string]any `json:"params"`
}

type fixtureCommandEntry struct {
	Run             bool           `json:"run"`
	NodeUnsupported bool           `json:"nodeUnsupported"`
	ExpectOk        bool           `json:"expectOk"`
	DataKeys        []string       `json:"dataKeys"`
	Command         fixtureCommand `json:"command"`
}

type meshFixture struct {
	Protocol         int                   `json:"protocol"`
	NodeCapabilities []string              `json:"nodeCapabilities"`
	Registration     map[string]any        `json:"registration"`
	ResultOk         map[string]any        `json:"resultOk"`
	ResultError      map[string]any        `json:"resultError"`
	SandboxFiles     map[string]string     `json:"sandboxFiles"`
	Commands         []fixtureCommandEntry `json:"commands"`
}

type eventsFixture struct {
	Protocol      int              `json:"protocol"`
	Events        []map[string]any `json:"events"`
	ForwardCompat []map[string]any `json:"forwardCompat"`
}

func loadFixture(t *testing.T, name string, out any) {
	t.Helper()
	raw, err := os.ReadFile(filepath.Join("..", "..", "protocol", "fixtures", name))
	if err != nil {
		t.Fatalf("read fixture %s: %v", name, err)
	}
	if err := json.Unmarshal(raw, out); err != nil {
		t.Fatalf("parse fixture %s: %v", name, err)
	}
}

func loadMeshFixture(t *testing.T) meshFixture {
	t.Helper()
	var fx meshFixture
	loadFixture(t, "mesh_v1.json", &fx)
	return fx
}

func TestNodeCapabilitiesMatchFixture(t *testing.T) {
	fx := loadMeshFixture(t)
	if !reflect.DeepEqual(fx.NodeCapabilities, nodeCapabilities) {
		t.Fatalf(
			"advertised capabilities drifted from protocol/fixtures/mesh_v1.json:\n fixture: %v\n    node: %v",
			fx.NodeCapabilities, nodeCapabilities,
		)
	}
}

func TestRegistrationBodyMatchesFixture(t *testing.T) {
	fx := loadMeshFixture(t)
	n := &Node{
		cfg:      &Config{Name: "rack-01"},
		DeviceID: "dev_node01",
	}
	body := n.registrationBody()

	// Same keys as the canonical registration — a renamed or dropped field
	// on either side is a silent registry miss, so key parity is the test.
	for key := range fx.Registration {
		if _, present := body[key]; !present {
			t.Errorf("registration body is missing fixture key %q", key)
		}
	}
	for key := range body {
		if _, present := fx.Registration[key]; !present {
			t.Errorf("registration body key %q is absent from the fixture", key)
		}
	}

	if body["capabilities"] == nil ||
		!reflect.DeepEqual(body["capabilities"], nodeCapabilities) {
		t.Errorf("registration must advertise nodeCapabilities, got %v", body["capabilities"])
	}
	platform, _ := body["platform"].(string)
	switch platform {
	case "linux", "macos", "windows":
	default:
		t.Errorf("platform %q is not a mesh DevicePlatform a node can register", platform)
	}
}

func TestResultBodyMatchesFixture(t *testing.T) {
	fx := loadMeshFixture(t)
	body := resultBody("dev_node01", commandResult{
		CommandID: "cmd_01",
		OK:        true,
		Message:   "done",
		Data:      map[string]any{"hostname": "rack-01"},
	})
	// Every key the canonical results use must exist in what the node posts.
	for key := range fx.ResultOk {
		if _, present := body[key]; !present {
			t.Errorf("result body is missing fixture key %q", key)
		}
	}
	for key := range fx.ResultError {
		if _, present := body[key]; !present {
			t.Errorf("result body is missing fixture key %q", key)
		}
	}
}

func TestDecodeCommandFrame(t *testing.T) {
	var fx eventsFixture
	loadFixture(t, "events_v1.json", &fx)

	frames := 0
	for _, event := range fx.Events {
		raw, err := json.Marshal(event)
		if err != nil {
			t.Fatal(err)
		}
		line := "data: " + string(raw)
		decoded, ok := decodeCommandFrame(line, "dev_node01")
		if event["kind"] == "device_command" && event["deviceId"] == "dev_node01" {
			frames++
			if !ok {
				t.Fatalf("canonical device_command frame was not decoded: %s", line)
			}
			if decoded["id"] != event["id"] || decoded["name"] != event["name"] {
				t.Fatalf("decoded frame lost fields: %v", decoded)
			}
			// The same frame addressed to a different node must be ignored.
			if _, misdelivered := decodeCommandFrame(line, "dev_other"); misdelivered {
				t.Fatal("frame for dev_node01 was accepted by dev_other")
			}
		} else if ok {
			t.Fatalf("non-command event %q must be skipped, got %v", event["kind"], decoded)
		}
	}
	if frames == 0 {
		t.Fatal("events fixture has no device_command sample addressed to dev_node01")
	}

	// Unknown kinds and junk must be tolerated, never fatal.
	for _, event := range fx.ForwardCompat {
		raw, _ := json.Marshal(event)
		if _, ok := decodeCommandFrame("data: "+string(raw), "dev_node01"); ok {
			t.Fatalf("forward-compat frame %v must be skipped", event)
		}
	}
	for _, line := range []string{"", ": keepalive", "data: {not json", "event: message"} {
		if _, ok := decodeCommandFrame(line, "dev_node01"); ok {
			t.Fatalf("malformed line %q must be skipped", line)
		}
	}
}

// TestDispatchConformance executes every runnable canonical command through
// the real dispatch() in a sandbox seeded from the fixture, asserting the
// outcome and the result-data keys the daemon's tools read.
func TestDispatchConformance(t *testing.T) {
	fx := loadMeshFixture(t)
	n := &Node{cfg: &Config{Name: "rack-01"}, DeviceID: "dev_node01"}

	for _, entry := range fx.Commands {
		if !entry.Run {
			continue
		}
		entry := entry
		t.Run(entry.Command.Name, func(t *testing.T) {
			sandbox := t.TempDir()
			for name, content := range fx.SandboxFiles {
				if err := os.WriteFile(
					filepath.Join(sandbox, name), []byte(content), 0o644,
				); err != nil {
					t.Fatal(err)
				}
			}
			params := make(map[string]any, len(entry.Command.Params))
			for k, v := range entry.Command.Params {
				if s, isString := v.(string); isString {
					params[k] = strings.ReplaceAll(s, "{{TMP}}", sandbox)
				} else {
					params[k] = v
				}
			}
			result := dispatch(context.Background(), n, entry.Command.Name, params)
			if result.OK != entry.ExpectOk {
				t.Fatalf(
					"dispatch(%s) ok=%v want %v (message: %s)",
					entry.Command.Name, result.OK, entry.ExpectOk, result.Message,
				)
			}
			for _, key := range entry.DataKeys {
				if _, present := result.Data[key]; !present {
					t.Errorf(
						"dispatch(%s) result data is missing contract key %q (got %v)",
						entry.Command.Name, key, result.Data,
					)
				}
			}
		})
	}
}

// TestDispatchUnsupportedAnswersCleanly pins the promise that a command
// outside the node's surface still gets an answer (never a timeout).
func TestDispatchUnsupportedAnswersCleanly(t *testing.T) {
	fx := loadMeshFixture(t)
	n := &Node{cfg: &Config{Name: "rack-01"}, DeviceID: "dev_node01"}
	tested := 0
	for _, entry := range fx.Commands {
		if !entry.NodeUnsupported {
			continue
		}
		tested++
		result := dispatch(context.Background(), n, entry.Command.Name, entry.Command.Params)
		if result.OK {
			t.Errorf("dispatch(%s) must fail on a node", entry.Command.Name)
		}
		if !strings.Contains(result.Message, "does not support") {
			t.Errorf("dispatch(%s) message %q should say the command is unsupported",
				entry.Command.Name, result.Message)
		}
	}
	if tested == 0 {
		t.Fatal("fixture has no nodeUnsupported command samples")
	}
}

File: src/__tests__/protocol-conformance.test.ts
/**
 * Daemon-side leg of the tri-implementation Bridge Protocol conformance
 * suite (see protocol/README.md). The same fixtures are replayed by the
 * companion app (protocol_conformance_test.dart) and talon-node
 * (protocol_conformance_test.go), so a wire drift on any side fails that
 * side's CI instead of shipping a silent misrender or a device command that
 * times out.
 *
 * The daemon is the producer of `BridgeEvent`s and the consumer of device
 * registrations/locations/command results, so this leg asserts:
 *   - the events fixture covers EVERY BridgeEvent kind (compile-time
 *     exhaustiveness: adding a kind without a fixture sample fails here);
 *   - each sample carries that kind's required fields;
 *   - mesh fixtures pass the real MeshRegistry validation/normalization;
 *   - the capability lists and command samples stay in lockstep with the
 *     device implementations' advertised surfaces.
 */

import { describe, it, expect } from "vitest";
import { readFileSync } from "node:fs";
import { mkdtemp } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";

import {
  BRIDGE_PROTOCOL_VERSION,
  type BridgeEvent,
  type DeviceCommand,
  type DeviceCommandResult,
  type DeviceInfo,
  type DeviceLocation,
} from "../frontend/native/protocol.js";
import { MeshRegistry } from "../core/mesh/index.js";
import {
  sanitizeCapabilities,
  toDeviceInfo,
  toDeviceLocation,
} from "../core/mesh/types.js";

const FIXTURES = join(__dirname, "../../protocol/fixtures");

const eventsFixture = JSON.parse(
  readFileSync(join(FIXTURES, "events_v1.json"), "utf-8"),
) as {
  protocol: number;
  events: BridgeEvent[];
  forwardCompat: Array<Record<string, unknown>>;
};

type CommandEntry = {
  run: boolean;
  nodeUnsupported?: boolean;
  expectOk: boolean;
  dataKeys: string[];
  command: DeviceCommand;
};

const meshFixture = JSON.parse(
  readFileSync(join(FIXTURES, "mesh_v1.json"), "utf-8"),
) as {
  protocol: number;
  nodeCapabilities: string[];
  companionCoreCapabilities: string[];
  companionDeviceControlCapabilities: string[];
  registration: Record<string, unknown>;
  companionRegistration: Record<string, unknown>;
  device: DeviceInfo;
  location: DeviceLocation;
  resultOk: DeviceCommandResult;
  resultError: DeviceCommandResult;
  sandboxFiles: Record<string, string>;
  commands: CommandEntry[];
};

/**
 * Every BridgeEvent kind, spelled out. `satisfies` rejects typos; the
 * AssertNever below fails to compile when a NEW kind is added to the union
 * without being listed here — which forces a fixture sample too, because the
 * runtime test asserts fixture kinds === this list.
 */
const EVENT_KINDS = [
  "hello",
  "status",
  "chat_created",
  "chat_updated",
  "chat_deleted",
  "message",
  "message_edited",
  "message_deleted",
  "reaction",
  "turn_start",
  "reasoning",
  "delta",
  "tool",
  "typing",
  "turn_end",
  "locate",
  "device_command",
  "error",
] as const satisfies readonly BridgeEvent["kind"][];

type MissingKind = Exclude<BridgeEvent["kind"], (typeof EVENT_KINDS)[number]>;
type AssertNever<T extends never> = T;
// eslint-disable-next-line @typescript-eslint/no-unused-vars
type _EveryKindListed = AssertNever<MissingKind>;

describe("events fixture (protocol/fixtures/events_v1.json)", () => {
  it("carries the daemon's protocol version", () => {
    expect(eventsFixture.protocol).toBe(BRIDGE_PROTOCOL_VERSION);
  });

  it("covers every BridgeEvent kind exactly", () => {
    const inFixture = new Set(eventsFixture.events.map((e) => e.kind));
    expect([...inFixture].sort()).toEqual([...EVENT_KINDS].sort());
  });

  it("every sample carries its kind's required fields", () => {
    for (const e of eventsFixture.events) {
      switch (e.kind) {
        case "hello":
          expect(e.status.app).toBe("talon-bridge");
          expect(e.status.protocol).toBe(BRIDGE_PROTOCOL_VERSION);
          expect(e.chats.length).toBeGreaterThan(0);
          expect(e.chats[0].id).toBeTypeOf("string");
          break;
        case "status":
          expect(e.status.botName).toBeTypeOf("string");
          expect(e.status.backend).toBeTypeOf("string");
          break;
        case "chat_created":
        case "chat_updated":
          expect(e.chat.id).toBeTypeOf("string");
          expect(e.chat.title).toBeTypeOf("string");
          expect(e.chat.createdAt).toBeTypeOf("number");
          expect(e.chat.lastActive).toBeTypeOf("number");
          expect(e.chat.preview).toBeTypeOf("string");
          break;
        case "chat_deleted":
          expect(e.chatId).toBeTypeOf("string");
          break;
        case "message":
          expect(e.chatId).toBe(e.message.chatId);
          expect(e.message.id).toBeTypeOf("string");
          expect(["user", "assistant", "system"]).toContain(e.message.role);
          expect(e.message.ts).toBeTypeOf("number");
          break;
        case "message_edited":
          expect(e.messageId).toBeTypeOf("string");
          expect(e.text).toBeTypeOf("string");
          break;
        case "message_deleted":
          expect(e.messageId).toBeTypeOf("string");
          break;
        case "reaction":
          expect(e.messageId).toBeTypeOf("string");
          expect(e.emoji.length).toBeGreaterThan(0);
          break;
        case "turn_start":
          expect(e.chatId).toBeTypeOf("string");
          break;
        case "reasoning":
        case "delta":
          expect(e.chatId).toBeTypeOf("string");
          expect(e.text).toBeTypeOf("string");
          break;
        case "tool":
          expect(e.id).toBeTypeOf("string");
          expect(e.name).toBeTypeOf("string");
          expect(["call", "result"]).toContain(e.phase);
          if (e.phase === "call") expect(e.input).toBeTypeOf("object");
          break;
        case "typing":
          expect(e.on).toBeTypeOf("boolean");
          break;
        case "turn_end":
          expect(e.delivered).toBeTypeOf("number");
          if (e.usage) {
            expect(e.usage.input).toBeTypeOf("number");
            expect(e.usage.output).toBeTypeOf("number");
          }
          break;
        case "locate":
          if (e.deviceId !== undefined) expect(e.deviceId).toBeTypeOf("string");
          break;
        case "device_command":
          expect(e.id).toBeTypeOf("string");
          expect(e.deviceId).toBeTypeOf("string");
          expect(e.name).toBeTypeOf("string");
          expect(e.params).toBeTypeOf("object");
          break;
        case "error":
          expect(e.message).toBeTypeOf("string");
          break;
        default: {
          const unreachable: never = e;
          throw new Error(`unhandled kind ${JSON.stringify(unreachable)}`);
        }
      }
    }
  });

  it("streamed tool phases agree with the persisted tool timeline", () => {
    // The assistant `message` sample persists the same tools the `tool`
    // events streamed — the two views of one turn must not drift.
    const streamed = eventsFixture.events.filter(
      (e): e is Extract<BridgeEvent, { kind: "tool" }> => e.kind === "tool",
    );
    const persisted = eventsFixture.events
      .filter(
        (e): e is Extract<BridgeEvent, { kind: "message" }> =>
          e.kind === "message",
      )
      .flatMap((e) => e.message.tools ?? []);
    for (const tool of persisted) {
      expect(
        streamed.some((s) => s.id === tool.id && s.name === tool.name),
      ).toBe(true);
    }
  });

  it("device_command samples use advertised capability names", () => {
    const known = new Set([
      ...meshFixture.nodeCapabilities,
      ...meshFixture.companionCoreCapabilities,
      ...meshFixture.companionDeviceControlCapabilities,
    ]);
    for (const e of eventsFixture.events) {
      if (e.kind === "device_command") expect(known).toContain(e.name);
    }
  });

  it("forward-compat frames stay parseable JSON objects with a kind", () => {
    expect(eventsFixture.forwardCompat.length).toBeGreaterThan(0);
    for (const frame of eventsFixture.forwardCompat) {
      expect(typeof frame.kind).toBe("string");
    }
  });
});

describe("mesh fixture (protocol/fixtures/mesh_v1.json)", () => {
  const freshRegistry = async () => {
    const dir = await mkdtemp(join(tmpdir(), "talon-conformance-"));
    return new MeshRegistry({
      devices: join(dir, "devices.json"),
      locations: join(dir, "locations.json"),
      history: join(dir, "history.json"),
    });
  };

  it("carries the daemon's protocol version", () => {
    expect(meshFixture.protocol).toBe(BRIDGE_PROTOCOL_VERSION);
  });

  it("node registration passes the real registry validation unchanged", async () => {
    const registry = await freshRegistry();
    const now = meshFixture.device.lastSeen;
    const device = await registry.register(meshFixture.registration, now);
    expect(device).toEqual(meshFixture.device);
  });

  it("companion registration passes the real registry validation", async () => {
    const registry = await freshRegistry();
    const device = await registry.register(
      meshFixture.companionRegistration,
      1767225600000,
    );
    expect(device.id).toBe("dev_pixel8");
    expect(device.platform).toBe("android");
    expect(device.battery).toBe(87);
    expect(device.charging).toBe(false);
    expect(device.capabilities).toEqual(meshFixture.companionCoreCapabilities);
  });

  it("location report passes the real registry validation unchanged", async () => {
    const registry = await freshRegistry();
    const stored = await registry.storeLocation(
      meshFixture.location as unknown as Record<string, unknown>,
      meshFixture.location.ts,
    );
    expect(stored).toEqual(meshFixture.location);
  });

  it("capability lists are canonical and within registry bounds", () => {
    for (const list of [
      meshFixture.nodeCapabilities,
      meshFixture.companionCoreCapabilities,
      meshFixture.companionDeviceControlCapabilities,
    ]) {
      // sanitizeCapabilities is what the daemon actually stores — canonical
      // lists must survive it byte-for-byte (lowercase, deduped, ≤16).
      expect(sanitizeCapabilities(list)).toEqual(list);
    }
  });

  it("node registration advertises exactly the canonical node capabilities", () => {
    expect(meshFixture.registration.capabilities).toEqual(
      meshFixture.nodeCapabilities,
    );
    expect(meshFixture.device.capabilities).toEqual(
      meshFixture.nodeCapabilities,
    );
  });

  it("every advertised capability has a canonical command sample", () => {
    const sampled = new Set(meshFixture.commands.map((c) => c.command.name));
    for (const name of [
      ...meshFixture.nodeCapabilities,
      ...meshFixture.companionCoreCapabilities,
      ...meshFixture.companionDeviceControlCapabilities,
    ]) {
      expect(
        sampled,
        `missing command sample for capability "${name}"`,
      ).toContain(name);
    }
    // And no sample invents a command outside the advertised surfaces.
    const known = new Set([
      ...meshFixture.nodeCapabilities,
      ...meshFixture.companionCoreCapabilities,
      ...meshFixture.companionDeviceControlCapabilities,
    ]);
    for (const name of sampled) expect(known).toContain(name);
  });

  it("commands marked node-unsupported are outside the node's surface", () => {
    for (const entry of meshFixture.commands) {
      const inNodeList = meshFixture.nodeCapabilities.includes(
        entry.command.name,
      );
      expect(inNodeList).toBe(!entry.nodeUnsupported);
    }
  });

  it("command samples carry correlation ids and JSON-object params", () => {
    for (const { command } of meshFixture.commands) {
      expect(command.id).toBeTypeOf("string");
      expect(command.id.length).toBeGreaterThan(0);
      expect(command.deviceId).toBeTypeOf("string");
      expect(command.params).toBeTypeOf("object");
    }
  });

  it("command-result samples match the wire contract", () => {
    const ok: DeviceCommandResult = meshFixture.resultOk;
    expect(ok.ok).toBe(true);
    expect(ok.commandId).toBeTypeOf("string");
    expect(ok.deviceId).toBeTypeOf("string");
    const err: DeviceCommandResult = meshFixture.resultError;
    expect(err.ok).toBe(false);
    expect(err.message).toBeTypeOf("string");
    expect(err.data).toBeUndefined();
  });

  it("normalizers are identity on the canonical device and location", () => {
    expect(
      toDeviceInfo(meshFixture.device, meshFixture.device.lastSeen),
    ).toEqual(meshFixture.device);
    expect(toDeviceLocation(meshFixture.location)).toEqual(
      meshFixture.location,
    );
  });
});
